cnd/src/db/save_message.rs

```diff
@@ -0,0 +1,353 @@
+use crate::{
+    db::{
+        custom_sql_types::{Text, U32},
+        new_types::{DecimalU256, EthereumAddress, Satoshis},
+        schema::{self, *},
+        Error, Sqlite,
+    },
+    swap_protocols::{
+        ledger::{Bitcoin, Ethereum},
+        rfc003::{Accept, Decline, Request, SecretHash},
+        HashFunction, SwapId,
+    },
+};
+use diesel::RunQueryDsl;
+
+pub trait SaveMessage<M> {
+    fn save_message(&self, message: M) -> Result<(), Error>;
+}
+
+pub trait SaveRfc003Messages:
+    SaveMessage<Request<Bitcoin, Ethereum, bitcoin::Amount, ethereum_support::EtherQuantity>>
+    + SaveMessage<Request<Bitcoin, Ethereum, bitcoin::Amount, ethereum_support::Erc20Token>>
+    + SaveMessage<Request<Ethereum, Bitcoin, ethereum_support::EtherQuantity, bitcoin::Amount>>
+    + SaveMessage<Request<Ethereum, Bitcoin, ethereum_support::Erc20Token, bitcoin::Amount>>
+    + SaveMessage<Accept<Bitcoin, Ethereum>>
+    + SaveMessage<Accept<Ethereum, Bitcoin>>
+    + SaveMessage<Decline>
+{
+}
+
+macro_rules! impl_save_message {
```

Comment:
📣 From the `Http<I>` experience I'd be keen to not use this macro.
It really does not give much (only function signatures and `let $conn = self.connect().unwrap();`) but decrease readability.

The next person who implement a new `save_message` will have to understand how to use the macro to then use it. I think it increases your mental load for little gain.

Reply:
Not having this macro makes you write all the types two times though :(
I would not consider this macro to be the long-term way of dealing with this duplication.

I see it as the starting point for iterating on a different way of dealing with this complexity in our codebase.

Thanks for pointing out the unwrap, that needs changing.